File: go/18_testing.go
package main

import "fmt"

func add(a, b int) int {
	return a + b
}

func divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, fmt.Errorf("cannot divide by zero")
	}
	return a / b, nil
}

func main() {
	// tiny self-check (real tests live in *_test.go with `go test`)
	if add(2, 3) != 5 {
		panic("add failed")
	}
	r, err := divide(10, 2)
	if err != nil || r != 5 {
		panic("divide failed")
	}
	_, err = divide(1, 0)
	if err == nil {
		panic("should error")
	}
	fmt.Println("all tests passed")
	fmt.Println("tip: put real tests in add_test.go and run: go test")
}

File: rust/18_testing.rs
fn add(a: i32, b: i32) -> i32 {
    a + b
}

fn divide(a: f64, b: f64) -> Result<f64, String> {
    if b == 0.0 {
        Err("cannot divide by zero".into())
    } else {
        Ok(a / b)
    }
}

fn main() {
    assert_eq!(add(2, 3), 5);
    assert_eq!(divide(10.0, 2.0).unwrap(), 5.0);
    assert!(divide(1.0, 0.0).is_err());
    println!("all tests passed");
    println!("tip: real tests use #[cfg(test)] mod tests {{ ... }}");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(add(2, 3), 5);
    }

    #[test]
    fn test_divide_ok() {
        assert_eq!(divide(10.0, 2.0).unwrap(), 5.0);
    }

    #[test]
    fn test_divide_err() {
        assert!(divide(1.0, 0.0).is_err());
    }
}
